File: android/app/src/main/java/com/arwheelapp/modules/OnnxOverlayView.kt
package com.arwheelapp.modules

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.view.View
import com.arwheelapp.utils.Detection

class OnnxOverlayView(context: Context) : View(context) {
    companion object {
        private const val TAG = "OnnxOverlayView"
        private const val INPUT_SIZE = 320f
    }

    private val detections: MutableList<Detection> = mutableListOf()

    private val boxPaint = Paint().apply {
        color = Color.CYAN
        style = Paint.Style.STROKE
        strokeWidth = 8f
    }

    private val textPaint = Paint().apply {
        color = Color.BLUE
        textSize = 40f
        style = Paint.Style.FILL
        isAntiAlias = true
    }

    fun updateDetections(newDetections: List<Detection>) {
        detections.clear()
        detections.addAll(newDetections)
        postInvalidate()
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        if (detections.isEmpty()) return

        val viewW = width.toFloat()
        val viewH = height.toFloat()
        val scale = maxOf(viewW, viewH) / INPUT_SIZE
        val offsetX = (viewW - (INPUT_SIZE * scale)) / 2
        val offsetY = (viewH - (INPUT_SIZE * scale)) / 2
        detections.forEach { det ->
            val bbox = det.boundingBox
            val left = (bbox.left * INPUT_SIZE * scale) + offsetX
            val top = (bbox.top * INPUT_SIZE * scale) + offsetY
            val right = (bbox.right * INPUT_SIZE * scale) + offsetX
            val bottom = (bbox.bottom * INPUT_SIZE * scale) + offsetY
            canvas.drawRect(left, top, right, bottom, boxPaint)
            canvas.drawText("${(det.confidence * 100).toInt()}%", left, top - 10, textPaint)
        }
    }

    fun clear() {
        detections.clear()
        invalidate()
    }
}

File: ios/ARWheelApp/Modules/OnnxOverlayView.swift
// Modules/OnnxOverlayView.swift
import UIKit

/// Transparent overlay view that draws detection bounding boxes.
/// Mirrors Android OnnxOverlayView exactly — same normalized-coord logic.
class OnnxOverlayView: UIView {
    private static let inputSize: CGFloat = 320

    private var detections: [Detection] = []
    private let boxColor = UIColor.cyan
    private let textColor = UIColor.blue

    // MARK: - Public

    func updateDetections(_ newDetections: [Detection]) {
        detections = newDetections
        DispatchQueue.main.async { self.setNeedsDisplay() }
    }

    func clear() {
        detections = []
        DispatchQueue.main.async { self.setNeedsDisplay() }
    }

    // MARK: - Drawing

    override func draw(_ rect: CGRect) {
        super.draw(rect)
        guard !detections.isEmpty, let ctx = UIGraphicsGetCurrentContext() else { return }

        let viewW = bounds.width
        let viewH = bounds.height

        let font = UIFont.systemFont(ofSize: 14)
        let textAttrs: [NSAttributedString.Key: Any] = [
            .font: font,
            .foregroundColor: textColor
        ]

        for det in detections {
            let b = det.boundingBox
            // boundingBox is normalized 0–1, convert to view coords
            let left   = b.minX * viewW
            let top    = b.minY * viewH
            let right  = b.maxX * viewW
            let bottom = b.maxY * viewH

            let boxRect = CGRect(x: left, y: top, width: right - left, height: bottom - top)

            ctx.setStrokeColor(boxColor.cgColor)
            ctx.setLineWidth(3)
            ctx.stroke(boxRect)

            let label = "\(Int(det.confidence * 100))%"
            let labelRect = CGRect(x: left, y: max(top - 20, 0), width: 50, height: 18)
            (label as NSString).draw(in: labelRect, withAttributes: textAttrs)
        }
    }
}
